File: Shared/ProviderSdk/Internal/TokenManager.cs
namespace ReportingPlatform.ProviderSdk.Internal;

internal sealed class TokenManager
{
    private readonly HttpClient _http;
    private readonly ProviderSdkOptions _opts;
    private readonly ILogger _logger;

    private string?        _cachedToken;
    private DateTimeOffset _issuedAt;
    private int            _expiresIn;  // seconds
    private int            _consecutive401s;

    // Sync root for cached token access
    private readonly SemaphoreSlim _lock = new(1, 1);

    // ActivitySource for telemetry
    private static readonly ActivitySource _actSource = ProviderSdkActivitySource.Source;

    public TokenManager(HttpClient http, ProviderSdkOptions opts, ILogger logger)
    {
        _http   = http;
        _opts   = opts;
        _logger = logger;
    }

    /// <summary>Returns cached token if still fresh (>90s remaining), else fetches a new one.</summary>
    public async Task<string> GetTokenAsync(CancellationToken ct)
    {
        await _lock.WaitAsync(ct);
        try
        {
            if (_cachedToken is not null && DateTimeOffset.UtcNow < _issuedAt.AddSeconds(_expiresIn - 90))
                return _cachedToken;
            return await FetchAndCacheAsync(ct);
        }
        finally { _lock.Release(); }
    }

    /// <summary>Force-acquire a fresh token (called by ConnectionManager on RefreshAuth / reconnect).</summary>
    public async Task<string> AcquireAsync(CancellationToken ct)
    {
        await _lock.WaitAsync(ct);
        try { return await FetchAndCacheAsync(ct); }
        finally { _lock.Release(); }
    }

    /// <summary>True when the cached token has passed the 80% lifetime threshold.</summary>
    public bool ShouldRefresh =>
        _cachedToken is not null &&
        DateTimeOffset.UtcNow > _issuedAt.AddSeconds(_expiresIn * _opts.TokenRefreshEarlyFraction);

    private async Task<string> FetchAndCacheAsync(CancellationToken ct)
    {
        // Must be called while _lock is held.
        using var activity = _actSource.StartActivity("sdk.token.acquire");

        var requestBody = new Models.TokenRequest
        {
            ClientId     = _opts.ClientId,
            ClientSecret = _opts.ClientSecret,
        };

        HttpResponseMessage response;
        try
        {
            response = await _http.PostAsJsonAsync(
                _opts.TokenEndpoint, requestBody,
                SdkJsonContext.Default.TokenRequest, ct);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Token endpoint network error");
            throw;
        }

        if (response.StatusCode == System.Net.HttpStatusCode.BadRequest)
        {
            var body = await response.Content.ReadAsStringAsync(ct);
            throw new ProviderSdkConfigurationException(
                $"Token endpoint returned 400 Bad Request — fix SDK configuration. Body: {body}");
        }

        if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
        {
            _consecutive401s++;
            _logger.LogWarning("Token endpoint 401 (consecutive={Count})", _consecutive401s);
            if (_consecutive401s >= 5)
            {
                _logger.LogCritical("5 consecutive 401s — credentials revoked or permanently invalid");
                throw new CredentialsRevokedException();
            }
            throw new HttpRequestException($"401 Unauthorized from token endpoint (attempt {_consecutive401s})");
        }

        if (response.StatusCode == System.Net.HttpStatusCode.TooManyRequests)
        {
            var retryAfterSec = response.Headers.RetryAfter?.Delta?.TotalSeconds ?? 60.0;
            _logger.LogWarning("Token endpoint 429 — RetryAfter={Seconds}s", retryAfterSec);
            // Rethrow — ConnectionManager will observe this in the backoff loop
            throw new HttpRequestException($"429 Too Many Requests — retry after {retryAfterSec}s")
            {
                Data = { ["retryAfterSeconds"] = retryAfterSec }
            };
        }

        response.EnsureSuccessStatusCode();
        _consecutive401s = 0; // reset on success

        var tokenResponse = await response.Content.ReadFromJsonAsync(
            SdkJsonContext.Default.TokenResponse, ct)
            ?? throw new InvalidOperationException("Token endpoint returned null response body.");

        _cachedToken = tokenResponse.AccessToken;
        _issuedAt    = DateTimeOffset.UtcNow;
        _expiresIn   = tokenResponse.ExpiresIn;

        _logger.LogDebug("Acquired JWT — expiresIn={ExpiresIn}s refreshAt={RefreshAt:O}",
            _expiresIn,
            _issuedAt.AddSeconds(_expiresIn * _opts.TokenRefreshEarlyFraction));

        return _cachedToken;
    }
}

File: samples/PythonProviderSample/token_manager.py
import time
import requests
import logging

logger = logging.getLogger(__name__)

class CredentialsRevokedException(Exception):
    pass

class TokenManager:
    def __init__(self, config):
        self._config = config
        self._token = None
        self._issued_at = 0.0
        self._expires_in = 0
        self._consecutive_401s = 0

    def get_token(self):
        """Returns cached token if >90s remaining; else fetches fresh."""
        if self._token and time.time() < self._issued_at + self._expires_in - 90:
            return self._token
        return self._fetch()

    def acquire(self):
        """Always fetches a fresh token."""
        return self._fetch()

    def should_refresh(self):
        """True when past 80% of token lifetime."""
        if not self._token:
            return False
        return time.time() > self._issued_at + self._expires_in * 0.80

    def _fetch(self):
        try:
            resp = requests.post(
                self._config.token_endpoint,
                json={
                    "clientId":     self._config.client_id,
                    "clientSecret": self._config.client_secret,
                    "grantType":    "client_credentials",
                },
                timeout=10,
            )
        except Exception as e:
            logger.warning("Token endpoint network error: %s", e)
            raise

        if resp.status_code == 400:
            raise RuntimeError(f"Token endpoint 400 Bad Request: {resp.text}")

        if resp.status_code == 401:
            self._consecutive_401s += 1
            logger.warning("Token endpoint 401 (consecutive=%d)", self._consecutive_401s)
            if self._consecutive_401s >= 5:
                logger.critical("5 consecutive 401s — credentials revoked")
                raise CredentialsRevokedException()
            raise RuntimeError(f"401 Unauthorized (attempt {self._consecutive_401s})")

        if resp.status_code == 429:
            retry_after = int(resp.headers.get("Retry-After", "60"))
            raise RuntimeError(f"429 Too Many Requests — retry after {retry_after}s")

        resp.raise_for_status()
        self._consecutive_401s = 0
        data = resp.json()
        self._token = data["accessToken"]
        self._issued_at = time.time()
        self._expires_in = data["expiresIn"]
        logger.debug("Acquired JWT — expiresIn=%ds", self._expires_in)
        return self._token
